### DISCUSSION TITLE
Where is type definitions of js-sdk?

### DISCUSSION THREAD

**Discussion by Mr-M1M3** [pocketbase/pocketbase#1652] [Q&A] [answered]
Hello, I need to use pocketbase with sveltekit. However, I'm not sure where to look for definitions. How do I get type definitions?

**Comment by ganigeorgiev** [comment, marked as answer]
I don't understand the question. The TS definitions are already shipped with the JS SDK - https://github.com/pocketbase/js-sdk/blob/master/package.json#L19.

If you want to manually import the types, you can refer to the definition files (`.d.ts` or `.d.mts`) located in the `dist` directory - https://github.com/pocketbase/js-sdk/tree/master/dist.

**Reply by Mr-M1M3** [reply]
I add pocketbase client to local variables in `handle()` hook like so:
```ts
export async function handle({event, resolve}){
// other code
event.locals.pb = new Pocketbase(env.PB);
//other code
}
```
The problem is that typescript warns me:
```
Property 'pb' does not exist on type 'Locals'.
```
In order to fix this, I have to add proper typing of local variable in `src/app.d.ts`
```ts
declare global {
	namespace App {
		interface Locals {
                    // add type of pb
		}
	}
}

export {};
```
But, I'm not sure how to import typescript type definition and use that type definition to add the type of pocketbase.

**Reply by ganigeorgiev** [reply]
You should be able to use directly the default import as shown in the [SvelteKit SSR example](https://github.com/pocketbase/js-sdk#ssr-integration), aka:

```ts
import PocketBase from "pocketbase";

declare global {
    declare namespace App {
        interface Locals {
            pb: PocketBase
        }
    }
}
```

**Reply by Mr-M1M3** [reply]
Can we add types to pocketbase response too?

**Reply by ganigeorgiev** [reply]
You can specify a custom type using generics:

```ts
pb.collection('tasks').getOne<Task>("RECORD_ID")
```

For more info you could check https://github.com/pocketbase/js-sdk#custom-record-types.

**Reply by Mr-M1M3** [reply]
I didn't mean this. Before I could return data from sveltekit load function, I have to ake the object serializable. To do so, I created a utlity function which accepts data returned from pocketbase. So, I need the type of the object that is returned by pb client. To conclude, I need `ListResult` in this case. 
```ts
Promise<ListResult<Task>>
```

**Reply by Mr-M1M3** [reply]
Like, I want to import type this way:
```typescript
import type {ListResult} from "the-type-definition-module-I-dont-know-but-wanted-to-know-all-this-time";
```

**Reply by ganigeorgiev** [reply]
`ListResult` is not a type but a JS class. To import it you can add it alongside your default `pocketbase` import, eg:

```js
import PocketBase, { ListResult } from "pocketbase";

// or just the ListResult
import { ListResult } from "pocketbase";
```

You can find all exported classes and types in https://github.com/pocketbase/js-sdk/blob/b4f503d5c5d9ba8f9769df44c14a330daaf7deab/src/index.ts#L32-L58.

**Reply by Mr-M1M3** [reply]
That's what I was looking for. Thanks ^_^. You have always been a great person.

**Reply by Mr-M1M3** [reply]
@ganigeorgiev Hello, sorry to bother again. How do I add types to `auth collections` ? 

And if you have extra time, can you tell what does this following statement import:
```typescript
const u = import('pocketbase' ).default['authStore']['model'];
```

**Reply by ganigeorgiev** [reply]
> How do I add types to auth collections?

I'm sorry, I don't understand what do you mean. Could you give an example?

If you asking for whether it is possible to specify a generic type for `pb.authStore.model`, then No, at the moment this is not possible since `model` is a union type of `Admin|Record|null`. As a hacky workaround, you can try with type assertion (eg. `pb.authStore.model as any as YourInterface`).


> And if you have extra time, can you tell what does this following statement import:
> `const u = import('pocketbase' ).default['authStore']['model'];`

Once again, you are not providing enough context to help you.
What are you trying to do and why do you need to import it this way? At first glance the above seems like a dynamic default import but technically it could be something else depending on your bundler configuration, whether you are using a esm or cjs modules and this is some kind of wrapper, etc.

**Reply by Mr-M1M3** [reply]
> If you asking for whether it is possible to specify a generic type for `pb.authStore.model`, 

Sort of. If it's not possible, at least we can do `authRefresh<UserType>(...)`, right?

> Once again, you are not providing enough context to help you. What are you trying to do and why do you need to import it this way?

Sorry about that :( . I saw someone to use this to add typings to sveltekit local variables : 
```ts
import Pocketbase from 'pocketbase';
declare global {
	namespace App {
		// interface Error {}
		interface Locals {
			pb: Pocketbase,
                       user: import('pocketbase' ).default['authStore']['model']
		}
		// interface PageData {}
		// interface Platform {}
	}
}

export {};

```

**Reply by avnav0** [reply]
@Mr-M1M3 i am running into same issue from same exact [source](https://www.youtube.com/watch?v=AxPB3e-3yEM&t=744s&ab_channel=JianYuanLee).  were you able to solve?

**Reply by Mr-M1M3** [reply]
@avnav88 not yet 🥲let me know if you find anything :)

**Reply by ganigeorgiev** [reply]
@Mr-M1M3, @avnav88 I'm confused. I thought that `pb: PocketBase` was the answer you were looking for, right?

Don't follow blindly tutorials. `import('pocketbase' ).default.*` in the above must be a "dynamic" import, but I don't see why you may need it and why it is necessary to have a separate `user` local prop.
If the goal of local `user` prop in the above is to hold a typed user structure, then you can assign to the interface your custom type, not the union one (_that may not even be defined, since I'm not sure how the above is resolved_).

If you have another question, please open a new discussion with more context, since this already has a lot of things going on and it is hard to follow and understand.

**Reply by avnav0** [reply]
i think the use was to have the user get authenticated after registration:
```
import type { Handle } from '@sveltejs/kit';
import { pb } from "./data/pocketbase";

console.log('pb-server:', pb);

export const handle: Handle = async ({ event, resolve}) => {
    //before
    pb.authstore.loadFromCookie( event.request.headers.get('cookie') )
    if (pb.authStore.isValid) {
        try {
            await pb.collection('users').authRefresh()
        } catch (_) {
            pb.authStore.clear();
        }
    }

//pass context to back-end handlers:
    event.locals.pb = pb;
    event.locals.user = structuredClone(pb.authStore.model)

    const response = await resolve(event);


    //after
    respose.headers.set('set-cookie', pb.authStore.exportToCookie({httpOnly: false}) )

    return response;
};
```
